orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,255 @@
+
+    auto& wait_output = graph.GetOrCreateNodeArg(new_name, &type_proto);
+    output_args.push_back(&wait_output);
+    input_args.push_back(old_input);
+    return wait_output;
+  };
+
+  if (recv_fw) {
+    // if we have recv op in forward pass (at the begining of the graph), add the WaitEvent op before that.
+    std::vector<NodeArg*> input_args;
+    std::vector<NodeArg*> output_args;
+    AddInputEvent(graph, "WaitEvent", true /* is_forward */, input_args, new_input_names);
+
+    // recv's first input is the signal input. Re-direct it to WaitEvent's input.
+    auto& input_signal = recv_fw->MutableInputDefs()[0];
+    auto& wait_output = update_wait_input_output(input_signal, input_args, output_args);
+    input_signal = &wait_output;
+
+    graph.AddNode(graph.GenerateNodeName("WaitEvent"),
+                  "WaitEvent",
+                  "",
+                  input_args,
+                  output_args,
+                  nullptr,
+                  kMSDomain);
+  } else {
+    // the first stage doesn't have recv_fw. Add Wait for all inputs.
+    std::vector<NodeArg*> input_args;
+    std::vector<NodeArg*> output_args;
+    AddInputEvent(graph, "WaitEvent", true /* is_forward */, input_args, new_input_names);
+    const std::vector<const NodeArg*>& graph_inputs = graph.GetInputs();
+
+    ORT_RETURN_IF_NOT(
+      graph_inputs.size() > 0, "Graph ", graph.Name(), " doesn't have any inputs.");
+
+    for (auto& input_arg : graph_inputs) {
+      NodeArg* mutable_input = graph.GetNodeArg(input_arg->Name());
+      auto& wait_output = update_wait_input_output(mutable_input, input_args, output_args);
+      std::vector<Node*> nodes = graph.GetMutableConsumerNodes(input_arg->Name());
+      for (auto& consumer_node : nodes) {
+        for (auto& i : consumer_node->MutableInputDefs()) {
+          if (i->Name() == input_arg->Name()) {
+            // if the node is fed by input, re-direct it to be fed by WaitEvent's output.
+            i = &wait_output;
+          }
+        }
+      }
+    }
+    graph.AddNode(graph.GenerateNodeName("WaitEvent"),
+                  "WaitEvent",
+                  "",
+                  input_args,
+                  output_args,
+                  nullptr,
+                  kMSDomain);
+  }
+  return Status::OK();
+}
+
+Status AddRecordForwardWaitBackward(Graph& graph, Node* send_fw, Node* recv_bw, std::vector<std::string>& new_input_names) {
+  ORT_RETURN_IF_NOT(
+      !send_fw == !recv_bw,
+      "Graph requires either having both send forward node "
+      "and recv backword node, or none of them. Currently the graph "
+      "has send forward: ",
+      send_fw, " and recv backward: ", recv_bw);
+
+  auto create_node_args = [&](const NodeArg* base_arg) -> NodeArg& {
+    const auto& new_name = graph.GenerateNodeArgName(base_arg->Name());
+    ONNX_NAMESPACE::TypeProto type_proto(*(base_arg->TypeAsProto()));
+    return graph.GetOrCreateNodeArg(new_name, &type_proto);
```

Comment:
Seems like we should throw if the node we want to create already exists.

Reply:
Good idea. I don't find a way to identify if the returned node is old or new, so added a check for existing node before creating. It would hurt the perf a little bit, but it is a one time thing on initialization so I guess it's OK.

---
In reply to: [409827827](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r409827827) [](ancestors = 409827827)